platforms/avr/drivers/ws2812_spi_int.c

```diff
@@ -0,0 +1,96 @@
+/*
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/* This is an interrupt-driven SPI driver for the
+ * WS2812/SK6812 RGB leds, for the ATmega32U.
+ *
+ * It sets SPI to 8MHz and sends 11100000 for a zero
+ * and 11111110 for a one. The time used by the ISR
+ * to prepare the next bit ensures the timming after
+ * each bit.
+ *
+ * It requires a 16MHz CPU clock and the pins PB0 and PB1
+ * unused (they are part of the ISP interface, so it's not
+ * uncommon), and the leds connected to the PB2 pin. Also,
+ * it can manage a maximum of 85 leds.
+ *
+ * It uses 40 bytes more than the WS2812 driver, but being interrupt
+ * based, allows to multiplex the LEDs with other tasks.
+ */
+
+#include <avr/interrupt.h>
+#include "rgblight.h"
+
+uint8_t *led_data;
+uint8_t  nbytes;
+uint8_t  remaining_bits;
+
+void ws2812_setleds(LED_TYPE *ledarray, uint16_t leds) {
+    led_data       = (uint8_t *)ledarray;
+    nbytes         = 3 * leds;
+    remaining_bits = 8;
+    // Initialize SPI for the WS2812/SK6812 driver
+    // Set PB2 and PB0 as output (not doing it freezes the SPI)
+    DDRB |= (1 << PB2);
+    DDRB |= (1 << PB0);
+    // Enable SPI, Master, set clock rate fck/2, MODE1, and enable interrupts
+    SPSR = (1 << SPI2X);
+    SPCR = (1 << SPIE) | (1 << SPE) | (1 << MSTR) | (1 << CPHA);
+    SPDR = 0; // send a byte to trigger an interrupt and begin the process
+}
```

Comment:
This code probably needs some protection against invocation when a transfer is already happening in the background, and enforcement of the `WS2812_TRST_US` delay between transfers (which I'm not sure how to do, unless some gross delay like 1…2 ms is used instead).